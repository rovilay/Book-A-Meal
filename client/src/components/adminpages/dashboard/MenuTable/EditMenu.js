/* eslint jsx-a11y/label-has-for:0 */
/* eslint max-len:0 */
import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import tableHead from '../../../../helpers/tableHead';
import TableHead from '../../../common/Table/TableHead';
import ModalTableRow from '../../../common/Table/ModalTableRow';

const EditMenuTable = (props) => {
  const {
    title,
    content,
    submitUpdate,
    editMenuMeals,
    serverRes,
    hideModal
  } = props;
  const sortedMeals = props.meals.sort((a, b) => a.title > b.title);
  const temp = [];
  const menuDate = moment(new Date(content.postOn)).format('DD/MM/YYYY');
  const editMenuMeal = (mealId) => {
    const checkbox = document.getElementById(`${mealId}-edit`);

    if (checkbox.checked === true) {
      props.addMealInEditMenu(mealId);
    } else if (checkbox.checked === false) {
      props.deleteMealInEditMenu(mealId);
    }
  };
  return (
    <div className="table-container">
      <h2 className="title">
        {title}
      </h2>

      <hr />
      <form onSubmit={(e) => {
        e.preventDefault();
        const meals = [...new Set(editMenuMeals.concat(temp))]; // merge new meals with old meals,returns only unique values
        submitUpdate(menuDate, meals);
        setTimeout(() => {
        }, 3000);
        if (serverRes.message === 'Menus updated successfully') {
          hideModal();
        }
      }}
      >
        <p className="big-little">
          Post On: {content.postOn}
        </p>
        <div className="edit-menu-meals">
          <h3 className="title-2">
            Add Meals
          </h3>
          <hr />
          <div className="checkbox-card">
            { sortedMeals.map(meal => (
              <p key={meal.id}>
                <input
                  type="checkbox"
                  name="meal-check"
                  className="meal-check"
                  id={`${meal.id}-edit`}
                  onClick={() => {
                    editMenuMeal(meal.id);
                  }}
                  value={meal.id}
                />
                {meal.title}
              </p>
            ))}
          </div>
        </div>
        <table>
          <TableHead tableHead={tableHead.editMenuHead} />
          <tbody className="modal-table-body">
            {
              content.meals.map((meal, i) => {
                const {
                  id,
                  title: Meal,
                  price: unitPrice,
                  description
                } = meal;
                temp.push(id);
                const item = {
                  sn: ++i,
                  Meal,
                  unitPrice,
                  description
                };

                return (
                  <ModalTableRow
                    key={id}
                    item={item}
                    id={id}
                    {...props}
                  />
                );
              })
            }
          </tbody>
        </table>

        <div className="menu-updatebtn">
          <button
            type="submit"
            name="updateMenuBtn"
            id="menu-btn"
            className="update-btn btn-1"
            disabled={temp.length <= 0}
          >
            Update Menu
          </button>
        </div>
      </form>
    </div>
  );
};

EditMenuTable.propTypes = {
  title: PropTypes.string.isRequired,
  content: PropTypes.object.isRequired,
  serverRes: PropTypes.object.isRequired,
  submitUpdate: PropTypes.func.isRequired,
  addMealInEditMenu: PropTypes.func.isRequired,
  deleteMealInEditMenu: PropTypes.func.isRequired,
  meals: PropTypes.array.isRequired,
  hideModal: PropTypes.func.isRequired,
  editMenuMeals: PropTypes.array.isRequired
};

export default EditMenuTable;
