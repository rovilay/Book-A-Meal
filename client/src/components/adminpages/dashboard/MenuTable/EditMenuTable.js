/* eslint jsx-a11y/label-has-for:0 */
/* eslint max-len:0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import sweetAlert from 'sweetalert';

class EditMenuTable extends Component {
  constructor(props) {
    super(props);

    this.getMealsNotInMenu = this.getMealsNotInMenu.bind(this);
    this.toggleMealCheckbox = this.toggleMealCheckbox.bind(this);
    this.updateMenu = this.updateMenu.bind(this);
  }

  /**
   * Checks meals for meals not currently in menu
   */
  getMealsNotInMenu() {
    const { content, meals } = this.props;
    const mealsInMenu = new Set(content.meals.map(meal => meal.id));
    return meals.filter(meal => !mealsInMenu.has(meal.id));
  }

  /**
   * check or unchecks meals
   * @param {*} meal meal to toggle
   */
  toggleMealCheckbox(meal) {
    return () => {
      const checkbox = document.getElementById(`${meal.id}-edit`);

      if (checkbox.checked === true) {
        this.props.addMealInEditMenu(meal.id);
      } else if (checkbox.checked === false) {
        this.props.deleteMealInEditMenu(meal.id);
      }
    };
  }

  /**
   * Updates menu
   * @param {*} event DOM event
   */
  /* istanbul ignore next */
  updateMenu(event) {
    event.preventDefault();
    const {
      hideModal,
      updateMenu,
      content,
      editMenuMeals
    } = this.props;
    const { menuId } = content;

    sweetAlert({
      text: 'Are you sure you want to update menu?',
      buttons: true,
      dangerMode: false,
    })
      .then((confirmed) => {
        if (confirmed) {
          const meals = [...editMenuMeals];
          updateMenu({ menuId, meals });
          hideModal();
        }
      })
      .catch(err => err);
  }

  render() {
    const {
      title,
      editMenuMeals,
    } = this.props;

    const mealsNotInMenu = this.getMealsNotInMenu();

    return (
      <div className="table-container">
        <h2 className="title">
          {title}
        </h2>

        <hr />
        <form
          onSubmit={this.updateMenu}
          className="editmenu-modal-form"
        >
          <div className="edit-menu-meals">
            <hr />
            {
              (mealsNotInMenu.length > 0)
                ?
                  <div className="checkbox-card">
                    {mealsNotInMenu.map(meal => (
                      <p key={meal.id} className="checkbox">
                        {meal.title}
                        <input
                          type="checkbox"
                          name="meal-check"
                          className="meal-check"
                          id={`${meal.id}-edit`}
                          onClick={this.toggleMealCheckbox(meal)}
                          value={meal.id}
                        />
                      </p>
                    ))}
                  </div>
                :
                  <div className="empty not-found">No meal to add!</div>
            }
          </div>

          {
            (mealsNotInMenu.length > 0)
            &&
            <div className="menu-updatebtn">
              <button
                type="submit"
                name="updateMenuBtn"
                id="menu-btn"
                className="btn-2"
                disabled={editMenuMeals.length <= 0}
              >
                Update Menu
              </button>
            </div>
          }
        </form>
      </div>
    );
  }
}

EditMenuTable.propTypes = {
  title: PropTypes.string.isRequired,
  addMealInEditMenu: PropTypes.func.isRequired,
  deleteMealInEditMenu: PropTypes.func.isRequired,
  meals: PropTypes.array.isRequired,
  hideModal: PropTypes.func.isRequired,
  editMenuMeals: PropTypes.array.isRequired,
  updateMenu: PropTypes.func.isRequired,
  content: PropTypes.object.isRequired
};

export default EditMenuTable;
