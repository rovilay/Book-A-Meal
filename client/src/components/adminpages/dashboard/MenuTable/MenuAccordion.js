import {
  AccordionItem,
  AccordionItemTitle,
  AccordionItemBody,
} from 'react-accessible-accordion';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import FontAwesome from 'react-fontawesome';
import moment from 'moment';
import sweetAlert from 'sweetalert';


class MenuAccordion extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isInfo: false,
    };

    this.handlePaginationClick = this.handlePaginationClick.bind(this);
    this.handleAccordionBodyClose = this.handleAccordionBodyClose.bind(this);
    this.handleAccordionItemTitleFocus = this.handleAccordionItemTitleFocus.bind(this);
    this.showMenuMeals = this.showMenuMeals.bind(this);
    this.hideMenuMeals = this.hideMenuMeals.bind(this);
    this.handleAddMealsToMenu = this.handleAddMealsToMenu.bind(this);
    this.deleteMealInMenu = this.deleteMealInMenu.bind(this);
  }

  /**
   * closes accordion body if menu meals is on show
   *
   * @param {object} className className of DOM element to watch for
   */
  handleAccordionBodyClose(className) {
    const element = document.querySelector(className);
    if (element) {
      const bodySelected = element.getAttribute('aria-selected');
      (bodySelected) && this.setState({ isInfo: false });
    }
  }

  /**
   * handles pagination changes
   *
   * @param {object} data data object from pagination component
   */
  handlePaginationClick(data) {
    const { Meals: mealUrl, } = this.props.item;
    const { limit } = this.props.menuMeals.pagination;
    const nextOffset = (data.selected) * limit;
    this.props.getMenuMeals(mealUrl, { limit, offset: nextOffset })
      .then(() => {
        this.setState({ isInfo: true });
      });
  }

  /**
   * closes accordion on focus
   */
  handleAccordionItemTitleFocus() {
    return this.handleAccordionBodyClose(`.myaccordiontitle-${this.props.item.menuId}`);
  }

  /**
   * shows menu meals
   * @param {*} event DOM event
   */
  showMenuMeals(event) {
    event.preventDefault();
    const { Meals: mealUrl } = this.props.item;
    this.props.getMenuMeals(mealUrl, {})
      .then(() => {
        this.setState({ isInfo: true });
      });
  }

  /**
   * hides menu meals
   * @param {*} event DOM event
   */
  hideMenuMeals(event) {
    event.preventDefault();
    this.setState({ isInfo: false });
  }

  /**
   * set menu to add meals
   * @param {*} event DOM event
   */
  handleAddMealsToMenu(event) {
    event.preventDefault();
    const { item, menuMeals } = this.props;
    const { meals } = menuMeals;

    const postOn = moment(item.postOn).format('LL');
    this.props.editMenu({ menuId: item.menuId, postOn, meals });
  }

  /**
   * deletes meal in a menu
   * @param {*} meal meal to delete
   */
  deleteMealInMenu(meal) {
    return (event) => {
      event.preventDefault();
      const { item, menuMeals } = this.props;
      const { Meals: mealUrl } = item;
      const { meals } = menuMeals;

      sweetAlert({
        text: 'Are you sure you want to remove this meal?',
        buttons: true,
        dangerMode: true,
      })
        .then((confirmed) => {
          if (confirmed) {
            const MenuMealsToDelete = meals.filter(n => n.id === meal.id);
            const MenuMealsToDeleteIds = MenuMealsToDelete.map(menuMeal => menuMeal.id);
            this.props.deleteMenuMeal({ mealUrl, meals: MenuMealsToDeleteIds });
          }
        })
        .catch(err => err);
    };
  }

  render() {
    const {
      isInfo,
    } = this.state;

    const {
      item,
      menuMeals,
    } = this.props;

    const {
      meals,
      pagination
    } = menuMeals;

    const {
      numOfPages,
      count
    } = pagination;

    const {
      sn,
      menuId,
      createdBy
    } = item;

    const postOn = moment(item.postOn).format('LL');
    const today = moment().format('YYYY-MM-DD');

    return (
      <AccordionItem
        key={sn}
        hiddenbodyclassname="myhide-accordion"
      >
        <AccordionItemTitle
          className={`accordion__title myaccordiontitle-${menuId}`}
          onFocus={this.handleAccordionItemTitleFocus}
        >
          <h3>
            <span className="serial">{sn}</span>
            <span className="postOn">{postOn}</span>
          </h3>
          <FontAwesome
            name="chevron-down"
          />
        </AccordionItemTitle>

        <AccordionItemBody>
          <div className="accordion-body">
            <div className="menuContainer">
              <div className="menu-details">
                <p className="menu-id">
                  <span>Post On:</span>
                  <span>{postOn}</span>
                </p>
                <p className="menu-id">
                  <span>Created By:</span>
                  <span>{createdBy}</span>
                </p>
              </div>

              <div className="menu-meals">
                {
                  (!isInfo)
                  &&
                  (
                  <button
                    className="btn-2 show-meals-btn"
                    onClick={this.showMenuMeals}
                  >
                    Show Meals
                  </button>
                  )
                }

                {
                  (isInfo) // for moble views
                  &&
                  (
                  <div className="meal-hide-add-btns show-mobile">
                    <button
                      className="responsive-btn-2 hide-meals-btn"
                      onClick={this.hideMenuMeals}
                    >
                      <FontAwesome
                        name = "eye-slash"
                        size = "lg"
                      />
                    </button>

                    {
                      // hide if menu date has passed
                      (moment(today).isSameOrBefore(item.postOn))
                      &&
                      <button
                        className="responsive-btn-2 add-meals-btn"
                        onClick={this.handleAddMealsToMenu}
                      >
                        <FontAwesome
                          name = "plus"
                          size = "lg"
                        />
                      </button>
                    }
                  </div>
                  )
                }
                {
                  (isInfo) // for large screen
                  &&
                  (
                  <div className="meal-hide-add-btns hide-mobile">
                    <button
                      className="responsive-btn-2 hide-meals-btn"
                      onClick={this.hideMenuMeals}
                    >
                      Hide Meals
                    </button>
                    {
                      // hide if menu date has passed
                      (moment(today, 'YYYY-MM-DD').isSameOrBefore(item.postOn))
                      &&
                      <button
                        className="responsive-btn-2 add-meals-btn"
                        onClick={this.handleAddMealsToMenu}
                      >
                        Add Meals
                      </button>
                    }
                  </div>
                  )
                }

                {
                  // menu meals
                  (meals.length > 0 && isInfo)
                  &&
                  <div className="meals">
                    {
                      meals.map((meal, i) => (
                        <div
                          id={meal.id}
                          className="meal"
                          key={i}
                        >
                          <span>
                            {meal.title}
                          </span>

                          {
                            // hide if menu date has passed
                            (moment(today, 'YYYY-MM-DD').isSameOrBefore(item.postOn))
                            &&
                            <button
                              className="btn-3 box-shadow deleteMeal"
                              onClick={this.deleteMealInMenu(meal)}
                            >
                              <FontAwesome
                                name="trash"
                                size="2x"
                                className="trash"
                              />
                            </button>
                          }
                        </div>
                      ))
                    }
                  </div>
                }

                {
                  (meals.length < 1 && isInfo)
                  &&
                  <span className="empty not-found">
                      Your meals are not in this menu
                  </span>
                }
              </div>
              {
                (count > 5 && isInfo)
                &&
                <div className="pagination-container">
                  <ReactPaginate
                    previousLabel="<<"
                    nextLabel=">>"
                    breakLabel={<a href="">...</a>}
                    breakClassName="break-me"
                    pageCount={numOfPages}
                    marginPagesDisplayed={2}
                    pageRangeDisplayed={5}
                    onPageChange={this.handlePaginationClick}
                    containerClassName="pagination"
                    subContainerClassName="pages pagination"
                    activeClassName="active"
                  />
                </div>
              }
            </div>
          </div>
        </AccordionItemBody>
      </AccordionItem>
    );
  }
}


MenuAccordion.propTypes = {
  item: PropTypes.object.isRequired,
  editMenu: PropTypes.func.isRequired,
  getMenuMeals: PropTypes.func.isRequired,
  deleteMenuMeal: PropTypes.func.isRequired,
  menuMeals: PropTypes.object.isRequired,
};

export default MenuAccordion;
