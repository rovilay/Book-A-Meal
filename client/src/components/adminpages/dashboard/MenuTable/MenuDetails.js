import React from 'react';
import PropTypes from 'prop-types';

import tableHeadData from '../../../../helpers/tableHeadData';
import TableHead from '../../../common/Table/TableHead';
import ModalTableRow from '../../../common/Table/ModalTableRow';

const MenuDetailsTable = (props) => {
  const { title, content } = props;
  const {
    menuDetails
  } = content;
  return (
    <div className="table-container">
      <h2 className="title">
        {title}
      </h2>
      <hr />
      <table>
        <TableHead tableHeadData={tableHeadData.menuDetailHead} />
        <tbody>
          {
            menuDetails.map((meal, i) => {
              const {
                id,
                title: Meal,
                price: unitPrice,
                description
              } = meal;
              const item = {
                sn: ++i,
                id,
                Meal,
                unitPrice,
                description
              };

              return (
                <ModalTableRow
                  key={i}
                  item={item}
                  sn={++i}
                  id={id}
                  {...props}
                />
              );
            })
          }
        </tbody>
      </table>
    </div>
  );
};

MenuDetailsTable.propTypes = {
  title: PropTypes.string.isRequired,
  content: PropTypes.object.isRequired,
};

export default MenuDetailsTable;
