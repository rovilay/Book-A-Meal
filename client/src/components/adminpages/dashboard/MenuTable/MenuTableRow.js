import React from 'react';
import PropTypes from 'prop-types';
import FontAwesome from 'react-fontawesome';

import TableCol from '../../../common/Table/TableCol';

const MenuTableRow = (props) => {
  const { item, showMenuDetails, editMenu } = props;
  const menu = {
    sn: item.sn,
    menuId: item.menuId,
    postOn: item.postOn,
    createdBy: item.createdBy,
  };

  return (
    <tr key={menu.menuId}>
      {
        Object.keys(menu).map((key, i) => (
          (<TableCol dataTitle={key} val={menu[key]} key={i} />)
        ))
      }

      {
        <td data-title="view details">
          <a
            role="button"
            href="#"
            onClick={() => {
              showMenuDetails(item.Meals);
            }}
            className="btn-col btn-2"
          >
            <FontAwesome
              name="info-circle"
              size="2x"
            />
          </a>
          <a
            onClick={() => {
              const { menuId, postOn, Meals: meals } = item;
              editMenu({ menuId, postOn, meals });
            }}
            href="#"
            role="button"
            className="btn-col btn-2"
          >
            <FontAwesome
              name="edit"
              size="2x"
            />
          </a>
        </td>
      }
    </tr>
  );
};

MenuTableRow.propTypes = {
  item: PropTypes.object.isRequired,
  showMenuDetails: PropTypes.func.isRequired,
  editMenu: PropTypes.func.isRequired
};

export default MenuTableRow;
