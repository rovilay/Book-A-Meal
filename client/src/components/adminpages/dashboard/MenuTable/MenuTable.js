import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import tableHead from '../../../../helpers/tableHead';
import TableHead from '../../../common/Table/TableHead';
import MenuTableRow from './MenuTableRow';

class MenuTable extends Component {
  render() {
    const sortedMenus = this.props.menus.sort((a, b) => new Date(b.postOn) - new Date(a.postOn));
    return (
      <div className="menu_table">
        <table>
          <TableHead tableHead={tableHead.menuTableHead} />
          <tbody>
            {
              sortedMenus.map((menu, i) => {
                const {
                  id: menuId,
                  postOn,
                  User,
                  Meals
                } = menu;
                const item = {
                  sn: ++i,
                  menuId,
                  postOn: moment(postOn).format('LL'),
                  createdBy: `${User.firstName} ${User.lastName}`,
                  Meals
                };
                return (<MenuTableRow
                  key={menuId}
                  item={item}
                  {...this.props}
                />);
              })
            }
          </tbody>
        </table>
      </div>
    );
  }
}

MenuTable.propTypes = {
  menus: PropTypes.array.isRequired,
};

export default MenuTable;
