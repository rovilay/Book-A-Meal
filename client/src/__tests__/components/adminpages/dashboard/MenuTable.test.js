import React from 'react';
import { shallow } from 'enzyme';
import MenuTable from '../../../../components/adminpages/dashboard/MenuTable/MenuTable';
import { menus } from '../../../__mockData__/menuMock';

describe('MenuTable component test', () => {
  const setup = () => {
    const props = {
      menus: [{
        ...menus[0],
        User: {
          firstName: 'ogooluwa',
          lastName: 'akinola'
        }
      }],
      getAllMenus: jest.fn(),
      pagination: {
        limit: 12,
        count: 13,
        numOfPages: 2
      },
      editMenu: jest.fn(),
      getMenuMeals: jest.fn()
    };

    return shallow(<MenuTable {...props} />)
  };


  it('should render MenuTable correctly', (done) => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();

    done();
  });


  it('should call `handlePaginationClick` if menu pagination button is clicked',
  (done) => {
    const wrapper = setup();

    const data = {
      selected: 1
    };

    const handlePaginationClickSpy= jest.spyOn(
      wrapper.instance(), 'handlePaginationClick'
    );
    wrapper.instance().handlePaginationClick(data)
    expect(handlePaginationClickSpy).toHaveBeenCalled();

    done();
  });
});
