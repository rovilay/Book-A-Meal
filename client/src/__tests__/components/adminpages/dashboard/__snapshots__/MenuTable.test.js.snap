// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuTable component test should render MenuTable correctly 1`] = `
<div
  className="accordion-container"
>
  <MenuAccordion
    editMenu={[MockFunction]}
    getAllMenus={[MockFunction]}
    getMenuMeals={[MockFunction]}
    item={
      Object {
        "Meals": undefined,
        "createdBy": "ogooluwa akinola",
        "menuId": "1",
        "postOn": "2018-09-05",
        "sn": NaN,
      }
    }
    key="1"
    menus={
      Array [
        Object {
          "User": Object {
            "firstName": "ogooluwa",
            "lastName": "akinola",
          },
          "id": "1",
          "meals": Array [],
          "postOn": "2018-09-05",
        },
      ]
    }
    pagination={
      Object {
        "count": 13,
        "limit": 12,
        "numOfPages": 2,
      }
    }
  />
  <div
    className="pagination-container"
  >
    <PaginationBoxView
      activeClassName="active"
      breakClassName="break-me"
      breakLabel={
        <a
          href=""
        >
          ...
        </a>
      }
      containerClassName="pagination"
      disableInitialCallback={false}
      disabledClassName="disabled"
      marginPagesDisplayed={2}
      nextClassName="next"
      nextLabel=">>"
      onPageChange={[Function]}
      pageCount={2}
      pageRangeDisplayed={5}
      previousClassName="previous"
      previousLabel="<<"
      subContainerClassName="pages pagination"
    />
  </div>
</div>
`;
